File: android/src/main/java/com/netrareactnative/dto/SlowNetworkPolicyActionDTO.kt
package com.example.netra_flutter.dto

import com.netra.library.enums.SlowNetworkPolicyAction
import kotlin.time.Duration.Companion.milliseconds

data class SlowNetworkPolicyActionDTO(
    val identifier: String,
    val delay: Double?,
    val timeout: Double?,
    val timeUnit: String?,
) {
    companion object {
        fun fromDataModel(slowNetworkPolicyAction: SlowNetworkPolicyAction): SlowNetworkPolicyActionDTO {
            return SlowNetworkPolicyActionDTO(
                identifier = slowNetworkPolicyAction.identifier,
                delay = ((if (slowNetworkPolicyAction is SlowNetworkPolicyAction.WAIT) {
                    slowNetworkPolicyAction.delay.inWholeMilliseconds
                } else {
                    null
                })?.toDouble()),
                timeout = ((if (slowNetworkPolicyAction is SlowNetworkPolicyAction.TIMEOUT) {
                    slowNetworkPolicyAction.timeout.inWholeMilliseconds
                } else {
                    null
                })?.toDouble()),
                timeUnit = ((if (slowNetworkPolicyAction is SlowNetworkPolicyAction.TIMEOUT) {
                    "MILLISECONDS"
                } else {
                    null
                }))
            )
        }
    }

    fun toDataModel(): SlowNetworkPolicyAction {
        return SlowNetworkPolicyAction.fromIdentifier(
            identifier,
            delay?.milliseconds,
            timeout?.milliseconds,
        )
    }
}

File: src/internal/dto/SlowNetworkPolicyActionDTO.ts
import { Duration } from '../../models';
import {
  SlowNetworkPolicyAction,
  TimeoutPolicyAction,
  WaitPolicyAction,
} from '../../models/SlowNetworkPolicyAction';
import { BaseDTO } from './BaseDTO';

export class SlowNetworkPolicyActionDTO extends BaseDTO {
  identifier: string;
  delay?: number;
  timeout?: number;
  timeUnit?: string;

  constructor(
    identifier: string,
    delay?: number,
    timeout?: number,
    timeUnit?: string
  ) {
    super();
    this.identifier = identifier;
    this.delay = delay;
    this.timeout = timeout;
    this.timeUnit = timeUnit;
  }

  static fromDataModel(model: SlowNetworkPolicyAction) {
    return new SlowNetworkPolicyActionDTO(
      model.identifier,
      model instanceof WaitPolicyAction
        ? model.delay.totalMilliseconds
        : undefined,
      model instanceof TimeoutPolicyAction
        ? model.timeout.totalMilliseconds
        : undefined,
      model instanceof TimeoutPolicyAction ? 'MILLISECONDS' : undefined
    );
  }

  static fromJSON(json: string) {
    const parsedJSON = JSON.parse(json);
    return new SlowNetworkPolicyActionDTO(
      parsedJSON.identifier,
      parsedJSON.delay,
      parsedJSON.timeout,
      parsedJSON.timeUnit
    );
  }

  toDataModel(): SlowNetworkPolicyAction {
    return SlowNetworkPolicyAction.fromIdentifier(
      this.identifier,
      this.delay !== undefined ? Duration.milliseconds(this.delay) : undefined,
      this.timeout !== undefined
        ? Duration.milliseconds(this.timeout)
        : undefined
    );
  }

  toJSON() {
    return {
      identifier: this.identifier,
      delay: this.delay,
      timeout: this.timeout,
      timeUnit: this.timeUnit,
    };
  }
}
